# Panle Regression - Firm Characteristics

### Random Effects Panel Regression

In [54]:
import pandas as pd
import numpy as np
import datetime as dt
import sklearn
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler
from linearmodels.panel import PanelOLS
from linearmodels.panel import RandomEffects


Data

In [55]:
df = pd.read_csv("Dataframes/characteristics.csv")
X_var_names = ["Revenue - Actual","Enterprise Value","Market Capitalization","loss firm status", "Enterprise Value To Sales (Daily Time Series Ratio)", "3 Month Total Return", "Volume"]
df

Instrument        Date GICS Industry Group Name  \
0          AVY.N  2013-01-01                Materials   
1          AVY.N  2013-04-01                Materials   
2          AVY.N  2013-07-01                Materials   
3          AVY.N  2013-10-01                Materials   
4          AVY.N  2014-01-01                Materials   
...          ...         ...                      ...   
19335    POOL.OQ  2021-10-01                Retailing   
19336    POOL.OQ  2022-01-01                Retailing   
19337    POOL.OQ  2022-04-01                Retailing   
19338    POOL.OQ  2022-07-01                Retailing   
19339    POOL.OQ  2022-10-01                Retailing   

       Earnings Per Share - Actual Surprise  Revenue - Actual  \
0                                    11.178      1.532200e+09   
1                                     2.482      1.498900e+09   
2                                     1.068      1.552300e+09   
3                                     8.095      1.504900e+09   
4                                     1.471      1.583900e+09   
...                                     ...               ...   
19335                                17.194      1.411448e+09   
19336                                40.267      1.035557e+09   
19337                                34.342      1.412650e+09   
19338                                 1.503      2.055818e+09   
19339                                 4.196      1.615339e+09   

       Enterprise Value  Market Capitalization  \
0          4.413001e+09           3.426001e+09   
1          5.459128e+09           4.309428e+09   
2          5.436229e+09           4.258229e+09   
3          5.029514e+09           4.273414e+09   
4          5.553068e+09           4.877168e+09   
...                 ...                    ...   
19335      1.769954e+10           1.742020e+10   
19336      2.384882e+10           2.268979e+10   
19337      1.843682e+10           1.696712e+10   
19338      1.556187e+10           1.405795e+10   
19339      1.406160e+10           1.259813e+10   

       Enterprise Value To Sales (Daily Time Series Ratio)  \
0                                               0.799347     
1                                               0.895787     
2                                               0.906078     
3                                               0.903176     
4                                               0.920659     
...                                                  ...     
19335                                           3.684400     
19336                                           4.504382     
19337                                           3.422880     
19338                                           2.749481     
19339                                           2.383926     

       3 Month Total Return    Volume  loss firm status  
0                 10.637479  207295.0                -1  
1                 24.148726       0.0                -1  
2                 -0.054218       0.0                -1  
3                  2.467620  236291.0                -1  
4                 16.003852  157761.0                -1  
...                     ...       ...               ...  
19335             -5.130322  157740.0                -1  
19336             30.489705   82349.0                -1  
19337            -25.160664  290897.0                -1  
19338            -21.661570  185584.0                -1  
19339             -9.164375  238539.0                -1  

[19340 rows x 11 columns]

In [56]:
df["Date"] = pd.to_datetime(df["Date"])

Remove outliers

In [57]:
#removing outliers
summary_stats = df["Earnings Per Share - Actual Surprise"].describe()
Q1 = summary_stats.loc['25%']
Q3 = summary_stats.loc['75%']
IQR = Q3 - Q1
threshold = 7
surprise_outliers_removed = df["Earnings Per Share - Actual Surprise"].loc[~((df["Earnings Per Share - Actual Surprise"] < (Q1 - threshold * IQR)) | (df["Earnings Per Share - Actual Surprise"] > (Q3 + threshold * IQR)))]
df_clean = df.copy()
df_clean["Earnings Per Share - Actual Surprise"] = surprise_outliers_removed
df_clean = df_clean.dropna(subset=["Earnings Per Share - Actual Surprise"])
#df_clean = df_clean.dropna()
df_panel = df_clean.copy()
df_clean

Instrument       Date GICS Industry Group Name  \
0          AVY.N 2013-01-01                Materials   
1          AVY.N 2013-04-01                Materials   
2          AVY.N 2013-07-01                Materials   
3          AVY.N 2013-10-01                Materials   
4          AVY.N 2014-01-01                Materials   
...          ...        ...                      ...   
19335    POOL.OQ 2021-10-01                Retailing   
19336    POOL.OQ 2022-01-01                Retailing   
19337    POOL.OQ 2022-04-01                Retailing   
19338    POOL.OQ 2022-07-01                Retailing   
19339    POOL.OQ 2022-10-01                Retailing   

       Earnings Per Share - Actual Surprise  Revenue - Actual  \
0                                    11.178      1.532200e+09   
1                                     2.482      1.498900e+09   
2                                     1.068      1.552300e+09   
3                                     8.095      1.504900e+09   
4                                     1.471      1.583900e+09   
...                                     ...               ...   
19335                                17.194      1.411448e+09   
19336                                40.267      1.035557e+09   
19337                                34.342      1.412650e+09   
19338                                 1.503      2.055818e+09   
19339                                 4.196      1.615339e+09   

       Enterprise Value  Market Capitalization  \
0          4.413001e+09           3.426001e+09   
1          5.459128e+09           4.309428e+09   
2          5.436229e+09           4.258229e+09   
3          5.029514e+09           4.273414e+09   
4          5.553068e+09           4.877168e+09   
...                 ...                    ...   
19335      1.769954e+10           1.742020e+10   
19336      2.384882e+10           2.268979e+10   
19337      1.843682e+10           1.696712e+10   
19338      1.556187e+10           1.405795e+10   
19339      1.406160e+10           1.259813e+10   

       Enterprise Value To Sales (Daily Time Series Ratio)  \
0                                               0.799347     
1                                               0.895787     
2                                               0.906078     
3                                               0.903176     
4                                               0.920659     
...                                                  ...     
19335                                           3.684400     
19336                                           4.504382     
19337                                           3.422880     
19338                                           2.749481     
19339                                           2.383926     

       3 Month Total Return    Volume  loss firm status  
0                 10.637479  207295.0                -1  
1                 24.148726       0.0                -1  
2                 -0.054218       0.0                -1  
3                  2.467620  236291.0                -1  
4                 16.003852  157761.0                -1  
...                     ...       ...               ...  
19335             -5.130322  157740.0                -1  
19336             30.489705   82349.0                -1  
19337            -25.160664  290897.0                -1  
19338            -21.661570  185584.0                -1  
19339             -9.164375  238539.0                -1  

[18569 rows x 11 columns]

#### Random Effects

In [58]:
# turn dataframe into right dimension for panel regression with multiindex
df_clean['Instrument'] = df_clean['Instrument'].astype('category')
df_clean['Date'] = pd.to_datetime(df_clean['Date'])

# set the index to be the time variable and the cross-sectional variable
df_clean.set_index(['Instrument', 'Date'], inplace=True)

# define the dependent variable



# perform the random effects panel regression
random_effects_model = RandomEffects(y, X)

# fit the model and print the summary statistics
random_effects_results = random_effects_model.fit()
print(random_effects_results.summary)


                                 RandomEffects Estimation Summary                                 
Dep. Variable:     Earnings Per Share - Actual Surprise   R-squared:                        0.0030
Estimator:                                RandomEffects   R-squared (Between):             -0.3708
No. Observations:                                 18569   R-squared (Within):               0.0112
Date:                                  Sun, Feb 26 2023   R-squared (Overall):             -0.0274
Time:                                          21:44:22   Log-likelihood                -7.829e+04
Cov. Estimator:                              Unadjusted                                           
                                                          F-statistic:                      9.1615
Entities:                                           501   P-value                           0.0000
Avg Obs:                                         37.064   Distribution:                 F(6,18562)
Min Obs:  

**R-squared (Within)**: This is the proportion of variation in the dependent variable (in your case, Earnings Per Share - Actual Surprise) that is explained by the independent variables (in our case, the variables in X) within each individual instrument (the 501 companies). In other words, this R-squared measures how well the independent variables explain the variation in the dependent variable (surprise) within each instrument. The R-squared (Within) is 0.0112, which means that the independent variables explain only about 1.12% of the variation in the dependent variable within each instrument.  

**R-squared (Between)**: This is the proportion of variation in the dependent variable that is explained by the independent variables across the instruments. In other words, this R-squared measures how well the independent variables explain the variation in the dependent variable between different instruments. The R-squared (Between) is -0.3708, which means that the independent variables actually perform worse than just using the mean of the dependent variable for each instrument to explain the variation in the dependent variable across entities.  

**R-squared (Overall)**: This is the proportion of variation in the dependent variable that is explained by the independent variables overall, taking into account both the within-instrument and between-instrument variation. The R-squared (Overall) is -0.0274, which means that the independent variables explain only about 2.74% of the variation in the dependent variable overall.

#### Fixed Effects

In [59]:
y = df_clean.loc[:, "Earnings Per Share - Actual Surprise"]
X = df_clean[X_var_names]

# perform the fixed effects panel regression
fixed_effects_model = PanelOLS(y, X, entity_effects=True, time_effects=True)

# fit the model and print the summary statistics
fixed_effects_results = fixed_effects_model.fit()
fixed_effects_results.summary


<class 'linearmodels.compat.statsmodels.Summary'>
"""
                                   PanelOLS Estimation Summary                                    
==================================================================================================
Dep. Variable:     Earnings Per Share - Actual Surprise   R-squared:                       -0.0029
Estimator:                                     PanelOLS   R-squared (Between):             -0.8660
No. Observations:                                 18569   R-squared (Within):               0.0056
Date:                                  Sun, Feb 26 2023   R-squared (Overall):             -0.0932
Time:                                          21:44:22   Log-likelihood                -7.773e+04
Cov. Estimator:                              Unadjusted                                           
                                                          F-statistic:                     -8.6700
Entities:                                           501   P-value                           1.0000
Avg Obs:                                         37.064   Distribution:                 F(6,18022)
Min Obs:                                         6.0000                                           
Max Obs:                                         40.000   F-statistic (robust):             482.92
                                                          P-value                           0.0000
Time periods:                                        41   Distribution:                 F(6,18022)
Avg Obs:                                         452.90                                           
Min Obs:                                         47.000                                           
Max Obs:                                         490.00                                           
                                                                                                  
                                                  Parameter Estimates                                                  
=======================================================================================================================
                                                     Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
-----------------------------------------------------------------------------------------------------------------------
Revenue - Actual                                     4.557e-10  3.635e-11     12.538     0.0000   3.844e-10   5.269e-10
Enterprise Value                                     1.661e-11  1.021e-11     1.6272     0.1037  -3.398e-12   3.662e-11
Market Capitalization                               -2.057e-11  1.034e-11    -1.9895     0.0467  -4.083e-11  -3.035e-13
loss firm status                                       -3.6511     0.2052    -17.794     0.0000     -4.0533     -3.2489
Enterprise Value To Sales (Daily Time Series Ratio)    -0.0093     0.0125    -0.7462     0.4556     -0.0339      0.0152
3 Month Total Return                                    0.1038     0.0093     11.160     0.0000      0.0855      0.1220
Volume                                                1.94e-07  5.415e-08     3.5819     0.0003   8.782e-08   3.001e-07
=======================================================================================================================

F-test for Poolability: 5.6197
P-value: 0.0000
Distribution: F(540,18022)

Included effects: Entity, Time
"""

By Industry: example tech industry

In [62]:
df2 = df_panel[df_panel['GICS Industry Group Name'] == 'Technology Hardware & Equipment']
groups = df2.groupby('Instrument')

def panel_regression_random(group):
    if len(group) < 2:
        return None
    
    group['Instrument'] = group['Instrument'].astype('category')
    group['Date'] = pd.to_datetime(group['Date'])
    group.set_index(['Instrument', 'Date'], inplace=True)

    y = group.loc[:, "Earnings Per Share - Actual Surprise"]
    X = group[X_var_names]

    model = RandomEffects(y, X).fit()
    return model


results_random = groups.apply(panel_regression_random)
results_random = results_random.dropna()
models_random = results_random.tolist()
models_random

[                                 RandomEffects Estimation Summary                                 
 Dep. Variable:     Earnings Per Share - Actual Surprise   R-squared:                        0.4285
 Estimator:                                RandomEffects   R-squared (Between):              0.0000
 No. Observations:                                    40   R-squared (Within):               0.4285
 Date:                                  Sun, Feb 26 2023   R-squared (Overall):              0.4285
 Time:                                          21:45:10   Log-likelihood                   -132.40
 Cov. Estimator:                              Unadjusted                                           
                                                           F-statistic:                      4.1245
 Entities:                                             1   P-value                           0.0034
 Avg Obs:                                         40.000   Distribution:                    F(6,33)
